src/java/org/apache/cassandra/repair/autorepair/AutoRepair.java

```diff
@@ -0,0 +1,401 @@
+    final static long INITIAL_REPAIR_DELAY_SEC = 30;
+
+    private static final Logger logger = LoggerFactory.getLogger(AutoRepair.class);
+
+    @VisibleForTesting
+    protected static Supplier<Long> timeFunc = System::currentTimeMillis;
+
+    public static AutoRepair instance = new AutoRepair();
+
+    @VisibleForTesting
+    protected final Map<AutoRepairConfig.RepairType, ScheduledExecutorPlus> repairExecutors;
+    @VisibleForTesting
+    protected final Map<AutoRepairConfig.RepairType, AutoRepairState> repairStates;
+
+    protected final Map<AutoRepairConfig.RepairType, IAutoRepairTokenRangeSplitter> tokenRangeSplitters = new EnumMap<>(AutoRepairConfig.RepairType.class);
+
+
+    @VisibleForTesting
+    protected AutoRepair()
+    {
+        AutoRepairConfig config = DatabaseDescriptor.getAutoRepairConfig();
+        repairExecutors = new EnumMap<>(AutoRepairConfig.RepairType.class);
+        repairStates = new EnumMap<>(AutoRepairConfig.RepairType.class);
+        for (AutoRepairConfig.RepairType repairType : AutoRepairConfig.RepairType.values())
+        {
+            repairExecutors.put(repairType, executorFactory().scheduled(false, "AutoRepair-Repair-" + repairType, Thread.NORM_PRIORITY));
+            repairStates.put(repairType, AutoRepairConfig.RepairType.getAutoRepairState(repairType));
+            tokenRangeSplitters.put(repairType, FBUtilities.newAutoRepairTokenRangeSplitter(config.getTokenRangeSplitter(repairType)));
+        }
+    }
+
+    public void setup()
+    {
+        verifyIsSafeToEnable();
+
+        AutoRepairConfig config = DatabaseDescriptor.getAutoRepairConfig();
+        AutoRepairService.setup();
+        AutoRepairUtils.setup();
+
+        for (AutoRepairConfig.RepairType repairType : AutoRepairConfig.RepairType.values())
+        {
+            repairExecutors.get(repairType).scheduleWithFixedDelay(
+            () -> repair(repairType, 60000),
+            INITIAL_REPAIR_DELAY_SEC,
+            config.getRepairCheckInterval().toSeconds(),
+            TimeUnit.SECONDS);
+        }
+    }
+
+    @VisibleForTesting
+    protected void verifyIsSafeToEnable()
+    {
+        AutoRepairConfig config = DatabaseDescriptor.getAutoRepairConfig();
+        if (config.isAutoRepairEnabled(AutoRepairConfig.RepairType.incremental) &&
+            (DatabaseDescriptor.getMaterializedViewsEnabled() || DatabaseDescriptor.isCDCEnabled()))
+            throw new ConfigurationException("Cannot enable incremental repair with materialized views or CDC enabled");
+    }
+
+    // repairAsync runs a repair session of the given type asynchronously.
+    public void repairAsync(AutoRepairConfig.RepairType repairType, long millisToWait)
+    {
+        repairExecutors.get(repairType).submit(() -> repair(repairType, millisToWait));
+    }
+
+    // repair runs a repair session of the given type synchronously.
+    public void repair(AutoRepairConfig.RepairType repairType, long millisToWait)
```

Comment:
NP: maybe simplify this method a little breaking it up, its very long and deep

Reply:
Chopped it a bit - see if this looks better now